File: library/src/main/java/com/imangazaliev/circlemenu/CenterMenuButton.java
package com.imangazaliev.circlemenu;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.animation.AnimatorSet;
import android.animation.ObjectAnimator;
import android.content.Context;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.os.Build;
import android.support.annotation.RequiresApi;
import android.support.v4.content.ContextCompat;
import android.view.animation.OvershootInterpolator;

public class CenterMenuButton extends CircleButton {

    private boolean isOpened;
    private AnimatorSet preLollipopAnimationSet;

    public  CenterMenuButton(Context context, boolean isOpened) {
        super(context);

        this.isOpened = isOpened;
        int colorNormal = getResources().getColor(R.color.circle_menu_center_button_color_normal);
        int colorPressed = getResources().getColor(R.color.circle_menu_center_button_color_pressed);

        setBackgroundCompat(createBackgroundDrawable(colorNormal, colorPressed));
        setImageResource(isOpened ? R.drawable.ic_close_vector : R.drawable.ic_menu_vector);

        //setVisibility(View.INVISIBLE);
    }

    public void setOpened(boolean isOpened) {
        this.isOpened = isOpened;
        if (isVectorAnimationSupported()) {
            startVectorAnimation(isOpened);
        } else {
            startPreLollipopAnimation();
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private void startVectorAnimation(boolean isOpened) {
        int iconId = isOpened ? R.drawable.ic_menu_animated : R.drawable.ic_close_animated;
        AnimatedVectorDrawable menuIcon = (AnimatedVectorDrawable) ContextCompat.getDrawable(getContext(), iconId);
        setImageDrawable(menuIcon);
        if (menuIcon != null) {
            menuIcon.start();
        }
    }

    private void startPreLollipopAnimation() {
        if (preLollipopAnimationSet == null) {
            preLollipopAnimationSet = createPreLollipopIconAnimation();
        }

        preLollipopAnimationSet.start();
    }

    private AnimatorSet createPreLollipopIconAnimation() {
        AnimatorSet preLollipopAnimationSet = new AnimatorSet();

        ObjectAnimator scaleOutX = ObjectAnimator.ofFloat(getDrawable(), "scaleX", 1.0f, 0.2f);
        ObjectAnimator scaleOutY = ObjectAnimator.ofFloat(getDrawable(), "scaleY", 1.0f, 0.2f);

        ObjectAnimator scaleInX = ObjectAnimator.ofFloat(this, "scaleX", 0.2f, 1.0f);
        ObjectAnimator scaleInY = ObjectAnimator.ofFloat(this, "scaleY", 0.2f, 1.0f);

        scaleOutX.setDuration(50);
        scaleOutY.setDuration(50);

        scaleInX.setDuration(150);
        scaleInY.setDuration(150);

        scaleInX.addListener(new AnimatorListenerAdapter() {
            @Override
            public void onAnimationStart(Animator animation) {
                setImageResource(isOpened ? R.drawable.ic_close_vector : R.drawable.ic_menu_vector);
            }
        });

        preLollipopAnimationSet.play(scaleOutX).with(scaleOutY);
        preLollipopAnimationSet.play(scaleInX).with(scaleInY).after(scaleOutX);
        preLollipopAnimationSet.setInterpolator(new OvershootInterpolator(2));
        return preLollipopAnimationSet;
    }

    private boolean isVectorAnimationSupported() {
        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP;
    }

}

File: library/src/main/java/com/imangazaliev/circlemenu/CenterMenuButton.kt
package com.imangazaliev.circlemenu

import android.animation.Animator
import android.animation.AnimatorListenerAdapter
import android.animation.AnimatorSet
import android.animation.ObjectAnimator
import android.annotation.SuppressLint
import android.content.Context
import android.content.res.ColorStateList
import android.graphics.drawable.AnimatedVectorDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.view.animation.OvershootInterpolator
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import com.google.android.material.floatingactionbutton.FloatingActionButton

@SuppressLint("ViewConstructor")
internal class CenterMenuButton(
        context: Context,
        backgroundColor: Int,
        menuIconType: CircleMenu.MenuIconType,
        private val iconColor: Int,
        private var isOpened: Boolean
) : FloatingActionButton(context) {

    private var preLollipopAnimationSet: AnimatorSet? = null
    private val menuIcon: MenuIcon by lazy {
        when (menuIconType) {
            CircleMenu.MenuIconType.HAMBURGER -> HamburgerMenuIcon()
            CircleMenu.MenuIconType.PLUS -> PlusMenuIcon()
        }
    }

    init {
        backgroundTintList = ColorStateList.valueOf(backgroundColor)
        setImageDrawable(getIconDrawable(!isOpened))
    }

    fun setOpened(isOpened: Boolean) {
        this.isOpened = isOpened
        if (isLollipop()) {
            startVectorAnimation(isOpened)
        } else {
            startPreLollipopAnimation()
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private fun startVectorAnimation(isOpening: Boolean) {
        val menuIcon = getIconDrawable(isOpening) as AnimatedVectorDrawable
        setImageDrawable(menuIcon)
        menuIcon.start()
    }

    private fun startPreLollipopAnimation() {
        if (preLollipopAnimationSet == null) {
            preLollipopAnimationSet = createPreLollipopIconAnimation()
        }
        preLollipopAnimationSet!!.start()
    }

    private fun createPreLollipopIconAnimation(): AnimatorSet {
        val preLollipopAnimationSet = AnimatorSet()
        val scaleOutX = ObjectAnimator.ofFloat(drawable, "scaleX", 1.0f, 0.2f)
        val scaleOutY = ObjectAnimator.ofFloat(drawable, "scaleY", 1.0f, 0.2f)
        val scaleInX = ObjectAnimator.ofFloat(this, "scaleX", 0.2f, 1.0f)
        val scaleInY = ObjectAnimator.ofFloat(this, "scaleY", 0.2f, 1.0f)
        scaleOutX.duration = 50
        scaleOutY.duration = 50
        scaleInX.duration = 150
        scaleInY.duration = 150
        scaleInX.addListener(object : AnimatorListenerAdapter() {
            override fun onAnimationStart(animation: Animator) {
                val icon = getIconDrawable(isOpened)
                setImageDrawable(icon)
            }
        })
        preLollipopAnimationSet.play(scaleOutX).with(scaleOutY)
        preLollipopAnimationSet.play(scaleInX).with(scaleInY).after(scaleOutX)
        preLollipopAnimationSet.interpolator = OvershootInterpolator(2f)
        return preLollipopAnimationSet
    }

    private fun getIconDrawable(isOpening: Boolean): Drawable {
        val iconResId = if (isLollipop()) {
            //animation from closed to opened
            if (isOpening) menuIcon.openingAnimatedIcon else menuIcon.closingAnimatedIcon
        } else {
            //if opening show close icon and vice versa
            if (isOpening) menuIcon.closeIcon else menuIcon.openIcon
        }

        val icon = ContextCompat.getDrawable(context, iconResId)!!
        icon.setTintCompat(iconColor)
        return icon
    }


}